src/synchronous_product/include/synchronous_product/synchronous_product.h

```diff
@@ -0,0 +1,209 @@
+ * triples (location, clock, valuation) (one for each clock), and then merged
+ * with the pairs from the ATAConfiguration. The resulting set is then
+ * partitioned according to the fractional part of the clock valuations. Then,
+ * each tuple is regionalized by replacing the clock valuation with the
+ * respective region index.  The resulting word is a sequence of sets, each set
+ * containing regionalized tuples that describe a TAState or ATAState. The
+ * sequence is sorted by the fractional part of the original clock valuations.
+ * @param ta_configuration The configuration of the timed automaton A
+ * @param ata_configuration The configuration of the alternating timed automaton B
+ * @param K The value of the largest constant any clock may be compared to
+ * @return The canonical word representing the state s, as a sorted vector of
+ * sets of tuples (triples from A and pairs from B).
+ */
+template <typename Location, typename ActionType>
+std::vector<std::set<ABRegionSymbol<Location, ActionType>>>
+get_canonical_word(const automata::ta::Configuration<Location> &ta_configuration,
+                   const ATAConfiguration<ActionType> &         ata_configuration,
+                   const unsigned int                           K)
+{
+	std::set<ABSymbol<Location, ActionType>> g;
+	// Insert ATA configurations into g.
+	std::copy(ata_configuration.begin(), ata_configuration.end(), std::inserter(g, g.end()));
+	// Insert TA configurations into g.
+	for (const auto &[clock_name, clock_value] : ta_configuration.second) {
+		g.insert(std::make_tuple(ta_configuration.first, clock_name, clock_value));
+	}
+	// Sort into partitions by the fractional parts.
+	std::map<double, std::set<ABSymbol<Location, ActionType>>> partitioned_g;
+	for (const ABSymbol<Location, ActionType> &symbol : g) {
+		partitioned_g[utilities::getFractionalPart<int, ClockValuation>(get_time(symbol))].insert(
+		  symbol);
+	}
+	// Replace exact clock values by region indices.
+	automata::ta::TimedAutomatonRegions                         regionSet{K};
+	std::vector<std::set<ABRegionSymbol<Location, ActionType>>> abs;
+	for (const auto &[fractional_part, g_i] : partitioned_g) {
+		std::set<ABRegionSymbol<Location, ActionType>> abs_i;
+		std::transform(
+		  g_i.begin(),
+		  g_i.end(),
+		  std::inserter(abs_i, abs_i.end()),
+		  [&](const ABSymbol<Location, ActionType> &w) -> ABRegionSymbol<Location, ActionType> {
+			  if (std::holds_alternative<TAState<Location>>(w)) {
+				  const TAState<Location> &s = std::get<TAState<Location>>(w);
+				  return TARegionState<Location>(std::get<0>(s),
+				                                 std::get<1>(s),
+				                                 regionSet.getRegionIndex(std::get<2>(s)));
+			  } else {
+				  const ATAState<ActionType> &s = std::get<ATAState<ActionType>>(w);
+				  return ATARegionState<ActionType>(s.first, regionSet.getRegionIndex(s.second));
+			  }
```

Comment:
The inner if-switch can be replaced by a visitor - do we want this? I will make one suggestion and you can check, whether you like it - if so, I can add further.

Reply:
Never mind, the visitor is overly verbose.